README.md
# student_elective_system_server

# 学生网上自主选课系统

> 服务端

## 系统简介

学生网上自主选课系统是一个前后端分离的学生选课系统, 项目采用 Egg.js + Vue.js 框架进行 开发。旨在实现一个学生能够在网上自主选课（选修课）的微系统。该系统分为三个角色：管理员, 学生, 教师；每一个角色的权限不同, 待操作项不同, 其中管理员能够对用户（学生、教师）和课程、专业、院系等信息进行增删改查；学生能进行选课、退选、查看成绩等操作；教师能够对学生的成绩进行评定, 能够查询到学生的选课状况等操作。

## 特色

- 前后端分离
- MVVM 模式、MVC 模式
- 前端采用最新的前端开发框架 Vue.js、Webpack 构建技术
- UI 框架采用 [Ant-Design-Vue](https://vuecomponent.github.io/ant-design-vue/docs/vue/introduce-cn/)
- 后端采用 egg.js 框架、Webpack 构建技术来构建一个 API 服务
- 跨域身份验证采用 JWT（JSON Web Token）
- API 接口规范遵守 RESTful 定义的规范
- 数据库采用 MySQL 数据库
- 采用 AJAX 异步请求方案
- 是一个 PC 端的 WEB 应用
- 分布式版本控制采用 Git
- 单元测试（目前只覆盖 GET 接口, PS：单元测试将临时关闭权限验证）

## 技术栈

- Vue.js
- Vuex.js
- Vue-Router.js
- axios.js
- Egg.js
- MySQL
- JSON Web Token
- Node.js
- Git
- Mockjs

## E-R

### E-R 完整版

![E-R_full](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R_full.png)

### E-R Mini 版

![E-R_mini](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R_mini.png)

## 系统功能结构图

### 系统功能结构图

![系统功能结构图](https://github.com/vxhly/egg-server/blob/master/.github/img/系统功能结构图.png)

## 用例图和数据流图

### ADMIN 权限用户登录之后能够进行的操作

`用例图`

![ADMIN](https://github.com/vxhly/egg-server/blob/master/.github/img/ADMIN.png)

`数据流图`

![ADMIN](https://github.com/vxhly/egg-server/blob/master/.github/img/ADMIN数据流图.png)


### STUDENT 权限用户登录之后能够进行的操作

`用例图`

![STUDENT](https://github.com/vxhly/egg-server/blob/master/.github/img/STUDENT.png)

`数据流图`

![STUDENT](https://github.com/vxhly/egg-server/blob/master/.github/img/STUDENT数据流图.png)

### TEACHER 权限用户登录之后能够进行的操作

`用例图`

![TEACHER](https://github.com/vxhly/egg-server/blob/master/.github/img/TEACHER.png)

`数据流图`

![TEACHER](https://github.com/vxhly/egg-server/blob/master/.github/img/TEACHER数据流图.png)

## 表结构

### 管理员表（admin)

![E-R-admin](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R-admin.png)

| 列名       | 类型         | 说明         | 备注                                                |
|------------|--------------|------------|-----------------------------------------------------|
| id         | int          | 自增长ID     | 主键, AUTO_INCREMENT                                |
| account    | varchar(20)  | 管理员账号   | 唯一值,长度5~20                                     |
| user_name  | varchar(10)  | 管理员用户名 | 长度5~10                                            |
| pwd        | varchar(255) | 登录密码     |                                                     |
| auth_token | text         | 登录令牌     | 允许空                                              |
| login_time | datetime     | 登录时间     | 允许空                                              |
| login_ip   | varchar(20)  | 登录 IP      | 允许空                                              |
| role       | varchar(10)  | 角色         | 默认值'ADMIN', 有效值：'ADMIN', 'STUDENT', 'TEACHER' |
| created_at | datetime     | 创建时间     |                                                     |
| updated_at | datetime     | 更新时间     |                                                     |
| deleted_at | datetime     | 删除时间     | 允许空                                              |

### 课程表（course）

![E-R-course](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R-course.png)

| 列名       | 类型         | 说明         | 备注                                                          |
|------------|--------------|------------|---------------------------------------------------------------|
| id         | int          | 自增长ID     | 主键, AUTO_INCREMENT                                          |
| c_id       | varchar(20)  | 课程编号     | 唯一值, 长度1~20                                              |
| t_id       | varchar(20)  | 授课教师工号 | 外键, 参考教师表的 t_id                                       |
| c_name     | varchar(30)  | 课程名称     | 长度2~30                                                      |
| c_info     | varchar(255) | 课程介绍     | 允许空, 默认''                                                |
| credit     | double       | 课程学分     | 默认值1.0, 有效值：0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5 |
| count      | int          | 可供选课人数 | 默认值10, 最小值0, 最大值100                                  |
| current    | int          | 当前被选人数 | 默认值0, 最小值0, 最大值100                                   |
| created_at | datetime     | 创建时间     |                                                               |
| updated_at | datetime     | 更新时间     |                                                               |
| deleted_at | datetime     | 删除时间     | 允许空                                                        |

### 院系表（dept）

![E-R-dept](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R-dept.png)

| 列名       | 类型         | 说明     | 备注                 |
|------------|--------------|--------|----------------------|
| id         | int          | 自增长ID | 主键, AUTO_INCREMENT |
| d_id       | varchar(20)  | 院系编号 | 唯一值, 长度1~20     |
| d_name     | varchar(30)  | 院系名称 | 长度2~30             |
| d_info     | varchar(255) | 院系介绍 | 允许空, 默认''       |
| created_at | datetime     | 创建时间 |                      |
| updated_at | datetime     | 更新时间 |                      |
| deleted_at | datetime     | 删除时间 | 允许空               |

### 专业表（major）

![E-R-major](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R-major.png)

| 列名       | 类型         | 说明     | 备注                    |
|------------|--------------|--------|-------------------------|
| id         | int          | 自增长ID | 主键, AUTO_INCREMENT    |
| m_id       | varchar(20)  | 专业编号 | 唯一值, 长度1~20        |
| d_id       | varchar(20)  | 院系编号 | 外键, 参照院系表的 d_id |
| m_name     | varchar(30)  | 专业名称 | 长度2~30                |
| m_info     | varchar(255) | 专业介绍 | 允许空, 默认''          |
| created_at | datetime     | 创建时间 |                         |
| updated_at | datetime     | 更新时间 |                         |
| deleted_at | datetime     | 删除时间 | 允许空                  |

### 学生表（student）

![E-R-student](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R-student.png)

| 列名       | 类型         | 说明     | 备注                                                  |
|------------|--------------|--------|-------------------------------------------------------|
| id         | int          | 自增长ID | 主键, AUTO_INCREMENT                                  |
| s_id       | varchar(20)  | 学号     | 唯一值, 长度5~20                                      |
| m_id       | varchar(20)  | 专业编号 | 外键, 参照专业表的 m_id                               |
| s_name     | varchar(5)   | 学生姓名 | 长度2~5                                               |
| sex        | varchar(5)   | 性别     | 默认'女', '男'或者'女'                                |
| pwd        | varchar(255) | 登录密码 |                                                       |
| phone      | varchar(20)  | 联系电话 | 默认''                                                |
| auth_token | text         | 登录令牌 | 允许空                                                |
| login_time | datetime     | 登录时间 | 允许空                                                |
| login_ip   | varchar(20)  | 登录 IP  | 允许空                                                |
| role       | varchar(10)  | 角色     | 默认值'STUDENT', 有效值：'ADMIN', 'STUDENT', 'TEACHER' |
| created_at | datetime     | 创建时间 |                                                       |
| updated_at | datetime     | 更新时间 |                                                       |
| deleted_at | datetime     | 删除时间 | 允许空                                                |

### 学生选课表（student_course）

![E-R-student_course](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R-student_course.png)

| 列名         | 类型        | 说明           | 备注                                      |
|--------------|-------------|--------------|-------------------------------------------|
| id           | int         | 自增长ID       | 主键, AUTO_INCREMENT                      |
| s_id         | varchar(20) | 学生学号       | 主键, 外键, 参照学生表的 s_id             |
| c_id         | varchar(20) | 课程编号       | 主键, 外键, 照课程表的 c_id               |
| score        | double      | 成绩           | 默认值0, 最小值0, 最大值100               |
| second_score | double      | 补考成绩       | 默认值0, 最小值0, 最大值100               |
| is_publish   | varchar(3)  | 是否已公布成绩 | 默认值'未公布', 有效值：'已公布', '未公布' |
| created_at   | datetime    | 创建时间       |                                           |
| updated_at   | datetime    | 更新时间       |                                           |
| deleted_at   | datetime    | 删除时间       | 允许空                                    |

### 教师表（teacher）

![E-R-teacher](https://github.com/vxhly/egg-server/blob/master/.github/img/E-R-teacher.png)

| 列名       | 类型         | 说明       | 备注                                                  |
|------------|--------------|----------|-------------------------------------------------------|
| id         | int          | 自增长ID   | 主键, AUTO_INCREMENT                                  |
| t_id       | varchar(20)  | 教师工号   | 唯一值, 长度5~20                                      |
| t_name     | varchar(5)   | 教师姓名   | 长度2~5                                               |
| sex        | varchar(5)   | 性别       | 默认'女', 有效值：'男', '女'                           |
| job        | varchar(5)   | 职称       | 默认'讲师', 有效值：'讲师', '教授', '副教授'           |
| count      | int          | 可供开课数 | 默认值10, 最小值0, 最大值20                           |
| current    | int          | 当前开课数 | 默认值0, 最小值0, 最大值20                            |
| pwd        | varchar(255) | 登录密码   |                                                       |
| phone      | varchar(20)  | 联系电话   | 默认''                                                |
| auth_token | text         | 登录令牌   | 允许空                                                |
| login_time | datetime     | 登录时间   | 允许空                                                |
| login_ip   | varchar(20)  | 登录 IP    | 允许空                                                |
| role       | varchar(10)  | 角色       | 默认值'TEACHER', 有效值：'ADMIN', 'STUDENT', 'TEACHER' |
| created_at | datetime     | 创建时间   |                                                       |
| updated_at | datetime     | 更新时间   |                                                       |
| deleted_at | datetime     | 删除时间   | 允许空                                                |

## QuickStart

<!-- add docs here for user -->

see [egg docs][egg] for more detail.

### 持续化部署

```bash
$ cd design-docker 
$ docker-compose up -d # 确保机子上有装 docker，第一次执行的时候将会下载容器会比较慢
$ cd ../
$ yarn install
$ yarn migrate
$ yarn dev
$ yarn test:data-user
$ yarn test:data
```

### testing

```bash
$ npm run test
```

### Deploy

```bash
$ npm start
$ npm stop
```

### npm scripts

- Use `npm run lint` to check code style.
- Use `npm test` to run unit test.
- Use `npm run autod` to auto detect dependencies upgrade, see [autod](https://www.npmjs.com/package/autod) for more detail.

[egg]: https://eggjs.org

CHANGELOG.md
## [0.2.2](https://github.com/vxhly/student_elective_system_server/compare/v0.2.1...v0.2.2) (2019-02-18)


### Features

* **swagger:** 添加 swagger 简易文档 ([cf88a53](https://github.com/vxhly/student_elective_system_server/commit/cf88a53))
* **swagger:** 添加swagger部分信息 ([97e2800](https://github.com/vxhly/student_elective_system_server/commit/97e2800))



## [0.2.1](https://github.com/vxhly/student_elective_system_server/compare/v0.2.0...v0.2.1) (2019-02-16)


### Bug Fixes

* **rule:** 修改422报错描述 ([f52fc41](https://github.com/vxhly/student_elective_system_server/commit/f52fc41))
* **service:** 修复部分数据返回错误 ([54c737f](https://github.com/vxhly/student_elective_system_server/commit/54c737f))



# [0.2.0](https://github.com/vxhly/student_elective_system_server/compare/81b373b...v0.2.0) (2019-02-14)


### Bug Fixes

* **app/controller/dept.js:** 修复部分问题 ([a4dd65a](https://github.com/vxhly/student_elective_system_server/commit/a4dd65a))
* **app/model/:** 修复模型的返回数据 ([126dd51](https://github.com/vxhly/student_elective_system_server/commit/126dd51))
* **controller/user.js:** 修复部分问题，包含用户信息的修改 ([de023b2](https://github.com/vxhly/student_elective_system_server/commit/de023b2))
* **status:** 修复错误 ([a929dbd](https://github.com/vxhly/student_elective_system_server/commit/a929dbd))
* **studentCourse:** 教师的账号只能查看自己的学生选修课情况，并且对其进行成绩的修改 ([bf5c988](https://github.com/vxhly/student_elective_system_server/commit/bf5c988))
* **studentCourse.add:** 修复选修的验证 ([cdd833e](https://github.com/vxhly/student_elective_system_server/commit/cdd833e))
* **StudentCourseController:** 修复部分错误 ([8974d89](https://github.com/vxhly/student_elective_system_server/commit/8974d89))
* **StudentCourseController.add:** 修复 ADMIN 不能修改发布状态bug ([639a721](https://github.com/vxhly/student_elective_system_server/commit/639a721))


### Features

* **/student-course/:** 添加课程的介绍 ([f07ec39](https://github.com/vxhly/student_elective_system_server/commit/f07ec39))
* **add student:** 新增一条学生信息的记录 ([85e7acd](https://github.com/vxhly/student_elective_system_server/commit/85e7acd))
* **admin AdminController.add AdminService.addAdmin:** 重写添加管理员方法 ([6c02373](https://github.com/vxhly/student_elective_system_server/commit/6c02373))
* **admin login:** 实现管理员登录 ([88d7c89](https://github.com/vxhly/student_elective_system_server/commit/88d7c89))
* **AdminController.getByAccount:** 添加获取管理员信息的接口，接口权限ADMIN ([8d34921](https://github.com/vxhly/student_elective_system_server/commit/8d34921))
* **AdminController.resetPwd:** 修改 admin 的密码 ([23f8cc7](https://github.com/vxhly/student_elective_system_server/commit/23f8cc7))
* **AdminController.update LoginController.index:** 完成 admin 登录验证,jwt 验证,admin 用户信息的更新 ([09bf8aa](https://github.com/vxhly/student_elective_system_server/commit/09bf8aa))
* **AdminService.findAllAdmin:** 查找所有的 admin 用户 ([b01490e](https://github.com/vxhly/student_elective_system_server/commit/b01490e))
* **app/controller/teacher.js:** 修复接口中的错误 ([54b441a](https://github.com/vxhly/student_elective_system_server/commit/54b441a))
* **bcrypt:** login 密码采用 bcrypt 加密算法 ([af3eb59](https://github.com/vxhly/student_elective_system_server/commit/af3eb59))
* **center -> destroy:** 各个管理中心的批量删除接口（实际上是修改 	deleted_at 的时间） ([d707570](https://github.com/vxhly/student_elective_system_server/commit/d707570))
* **check_rule.checkRule:** 接口验证权限中间件 ([7a079e6](https://github.com/vxhly/student_elective_system_server/commit/7a079e6))
* **config(mysql):** 配置 MySQL,生成建库建表 sql 语句 ([81b373b](https://github.com/vxhly/student_elective_system_server/commit/81b373b))
* **controller/admin:** 添加 admin ,统一错误处理, 跨域, ([fd04f72](https://github.com/vxhly/student_elective_system_server/commit/fd04f72))
* **CourseController.add:** 教师开课数的验证 ([88e97ed](https://github.com/vxhly/student_elective_system_server/commit/88e97ed))
* **CourseController.all model.Course CourseService.findAllCourse:** 查询所有的课程 ([d08ee5a](https://github.com/vxhly/student_elective_system_server/commit/d08ee5a))
* **CourseController.getById DeptController.getById:** 根据课程编号获取课程信息,根据院系编号获取院系信息 ([8e2888e](https://github.com/vxhly/student_elective_system_server/commit/8e2888e))
* **CourseController.update:** 更新课程信息 ([0e5315f](https://github.com/vxhly/student_elective_system_server/commit/0e5315f))
* **DeptController.add model.Dept DeptService.addDept:** 添加一条院系信息 ([1fc908d](https://github.com/vxhly/student_elective_system_server/commit/1fc908d))
* **DeptController.all:** 获取所有的院系信息添加url请求参数 name ([e3ad649](https://github.com/vxhly/student_elective_system_server/commit/e3ad649))
* **DeptController.delete:** 批量删除 ([47d3d96](https://github.com/vxhly/student_elective_system_server/commit/47d3d96))
* **DeptController.update AdminService.findAllAdmin DeptService.findAllDept:** 更新院系信息,, admin\dept 分 ([43abac2](https://github.com/vxhly/student_elective_system_server/commit/43abac2))
* **DeptService.findAllDept AdminService.findAllAdmin:** 查询所有的院系,修复分页 bug ([d60af35](https://github.com/vxhly/student_elective_system_server/commit/d60af35))
* **error_handler.errorHandler:** 404 拦截返回信息 ([42bad89](https://github.com/vxhly/student_elective_system_server/commit/42bad89))
* **helper verifyToken:** 更严格的接口权限验证,解决重新登录的时候未过期的token还是可以用以接口访问 ([3cab867](https://github.com/vxhly/student_elective_system_server/commit/3cab867))
* **helper.setScore:** 学生账号不能看未发布的成绩，将其设置成 0 ([16ff3d2](https://github.com/vxhly/student_elective_system_server/commit/16ff3d2))
* **jwt:** json wen token ([7b8b822](https://github.com/vxhly/student_elective_system_server/commit/7b8b822))
* **MajorController.all:** 获取所有的专业信息添加url请求参数 name ([4b303e0](https://github.com/vxhly/student_elective_system_server/commit/4b303e0))
* **MajorService.findAllMajor:** 关联查询,关联表 dept ([7040279](https://github.com/vxhly/student_elective_system_server/commit/7040279))
* **model.Admin:** 修改字段,相应的代码进行修改 ([7204895](https://github.com/vxhly/student_elective_system_server/commit/7204895))

app/utils/rule.js
'use strict';

module.exports = {
  AdminRule: {
    account: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '管理员账号不能为空',
      min: 5,
      max: 20,
    },
    user_name: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '管理员用户名不能为空',
      min: 2,
      max: 10,
    },
    pwd: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '管理员密码不能为空',
    },
    role: {
      type: 'enum',
      required: false,
      allowEmpty: true,
      values: [ 'ADMIN', 'STUDENT', 'TEACHER' ],
    },
  },

  DeptRule: {
    d_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '院系编号不能为空,而且必须为数字,首位不能为0',
      min: 1,
      max: 20,
    },
    d_name: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '院系名称不能为空',
      min: 2,
      max: 30,
    },
    d_info: {
      type: 'string',
      required: false,
      allowEmpty: true,
      min: 0,
      max: 255,
    },
  },

  MajorRule: {
    m_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '专业编号不能为空,而且必须为数字,首位不能为0',
      min: 1,
      max: 20,
    },
    d_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '院系编号不能为空,而且必须为数字,首位不能为0',
      min: 1,
      max: 20,
    },
    m_info: {
      type: 'string',
      required: false,
      allowEmpty: true,
      min: 0,
      max: 255,
    },
  },

  StudentRule: {
    s_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '学号不能为空,且必须为数字,首位不能为0',
      min: 5,
      max: 20,
    },
    m_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '专业编号不能为空,而且必须为数字,首位不能为0',
      min: 1,
      max: 20,
    },
    s_name: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '姓名不能为空',
      min: 2,
      max: 5,
    },
    sex: {
      type: 'enum',
      required: true,
      allowEmpty: false,
      values: [ '男', '女' ],
    },
    phone: {
      type: 'string',
      required: false,
      allowEmpty: true,
      format: /^(13[0-9]|14[579]|15([0-3]|[5-9])|16[6]|17[0135678]|18[0-9]|19[89])\d{8}$/,
      message: '不是正确格式的电话号码',
    },
    role: {
      type: 'enum',
      required: false,
      allowEmpty: true,
      values: [ 'ADMIN', 'STUDENT', 'TEACHER' ],
    },
  },

  TeacherRule: {
    t_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '教师工号不能为空,且必须为数字,首位不能为0',
      min: 5,
      max: 20,
    },
    t_name: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '姓名不能为空',
      min: 2,
      max: 5,
    },
    sex: {
      type: 'enum',
      required: true,
      allowEmpty: false,
      values: [ '男', '女' ],
    },
    job: {
      type: 'enum',
      required: false,
      allowEmpty: true,
      values: [ '讲师', '教授', '副教授' ],
    },
    count: {
      type: 'number',
      required: true,
      allowEmpty: false,
      min: 0,
      max: 20,
    },
    phone: {
      type: 'string',
      required: false,
      allowEmpty: true,
      format: /^(13[0-9]|14[579]|15([0-3]|[5-9])|16[6]|17[0135678]|18[0-9]|19[89])\d{8}$/,
      message: '不是正确格式的电话号码',
    },
    role: {
      type: 'enum',
      required: false,
      allowEmpty: true,
      values: [ 'ADMIN', 'STUDENT', 'TEACHER' ],
    },
  },

  StudentCourseRule: {
    s_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '学号不能为空,且必须为数字,首位不能为0',
      min: 5,
      max: 20,
    },
    c_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '课程编号不能为空,而且必须为数字,首位不能为0',
      min: 1,
      max: 20,
    },
    score: {
      type: 'number',
      required: false,
      allowEmpty: false,
      min: 0,
      max: 100,
    },
    second_score: {
      type: 'number',
      required: false,
      allowEmpty: false,
      min: 0,
      max: 100,
    },
    is_publish: {
      type: 'string',
      required: false,
      allowEmpty: true,
      values: [ '未公布', '已公布' ],
    },
  },

  CourseRule: {
    c_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '课程编号不能为空,而且必须为数字,首位不能为0',
      min: 1,
      max: 20,
    },
    t_id: {
      type: 'string',
      required: true,
      allowEmpty: false,
      format: /^[1-9]\d*$/,
      message: '课程编号不能为空,而且必须为数字,首位不能为0',
      min: 1,
      max: 20,
    },
    c_name: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '课程名称不能为空',
      min: 2,
      max: 30,
    },
    c_info: {
      type: 'string',
      required: false,
      allowEmpty: true,
      min: 0,
      max: 255,
    },
    credit: {
      type: 'enum',
      required: true,
      allowEmpty: false,
      min: 0.5,
      max: 4.5,
      values: [ 0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5 ],
    },
    count: {
      type: 'number',
      required: true,
      allowEmpty: false,
      min: 0,
      max: 100,
    },
  },

  DeleteRule: {
    id: {
      type: 'array',
      required: true,
      allowEmpty: false,
    },
  },

  LoginRule: {
    account: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '用户账号不能为空',
      min: 5,
      max: 20,
    },
    pwd: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '用户密码不能为空',
    },
    role: {
      type: 'enum',
      required: true,
      allowEmpty: false,
      values: [ 'ADMIN', 'STUDENT', 'TEACHER' ],
    },
    rememberMe: {
      type: 'boolean',
      required: false,
      allowEmpty: false,
      defaultValue: false,
    },
  },

  PwdRule: {
    old_password: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '旧密码不能为空',
    },
    new_password: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '新密码不能为空',
    },
    repeat_password: {
      type: 'string',
      required: true,
      allowEmpty: false,
      message: '再次输入的密码不能为空',
    },
  },

};
